Flaky test: Test/AddressesRemoved

https://github.com/grpc/grpc-go/pull/4737/checks?check_run_id=3536496494

```
--- FAIL: Test (1.03s)
    --- FAIL: Test/AddressesRemoved (0.52s)
        tlogger.go:101: clientconn.go:252 [core] parsed scheme: "whatever"  (t=+549.531µs)
        tlogger.go:101: resolver_conn_wrapper.go:100 [core] ccResolverWrapper: sending update to cc: {[{127.0.0.1:35011  <nil> 0 <nil>}] <nil> <nil>}  (t=+2.127622ms)
        tlogger.go:101: clientconn.go:1142 [core] Subchannel Connectivity change to CONNECTING  (t=+6.468872ms)
        tlogger.go:101: clientconn.go:1253 [core] Subchannel picks a new address "127.0.0.1:35011" to connect  (t=+6.573078ms)
        tlogger.go:101: clientconn.go:436 [core] Channel Connectivity change to CONNECTING  (t=+6.925999ms)
        tlogger.go:101: clientconn.go:1142 [core] Subchannel Connectivity change to READY  (t=+8.460187ms)
        tlogger.go:101: roundrobin.go:50 [roundrobin] roundrobinPicker: newPicker called with info: {map[0x40000cd660:{{127.0.0.1:35011  <nil> 0 <nil>}}]}  (t=+8.618796ms)
        tlogger.go:101: clientconn.go:436 [core] Channel Connectivity change to READY  (t=+8.698501ms)
        tlogger.go:101: resolver_conn_wrapper.go:100 [core] ccResolverWrapper: sending update to cc: {[] <nil> <nil>}  (t=+13.72159ms)
        tlogger.go:101: clientconn.go:436 [core] Channel Connectivity change to TRANSIENT_FAILURE  (t=+13.881099ms)
        tlogger.go:101: clientconn.go:1142 [core] Subchannel Connectivity change to SHUTDOWN  (t=+14.152615ms)
        tlogger.go:101: roundrobin.go:50 [roundrobin] roundrobinPicker: newPicker called with info: {map[]}  (t=+14.385528ms)
        tlogger.go:101: clientconn.go:436 [core] Channel Connectivity change to IDLE  (t=+14.456532ms)
        roundrobin_test.go:245: timed out waiting for state change.  got IDLE; want TRANSIENT_FAILURE
        tlogger.go:101: clientconn.go:436 [core] Channel Connectivity change to SHUTDOWN  (t=+515.883508ms)
FAIL
FAIL	google.golang.org/grpc/balancer/roundrobin	2.115s
```

--- a/balancer/balancer.go
+++ b/balancer/balancer.go
@@ -372,6 +372,7 @@ type ConnectivityStateEvaluator struct {
 	numReady            uint64 // Number of addrConns in ready state.
 	numConnecting       uint64 // Number of addrConns in connecting state.
 	numTransientFailure uint64 // Number of addrConns in transient failure state.
+	numIdle             uint64 // Number of addrConns in idle state.
 }
 
 // RecordTransition records state change happening in subConn and based on that
@@ -380,7 +381,8 @@ type ConnectivityStateEvaluator struct {
 //  - If at least one SubConn in Ready, the aggregated state is Ready;
 //  - Else if at least one SubConn in Connecting, the aggregated state is Connecting;
 //  - Else if at least one SubConn is TransientFailure, the aggregated state is Transient Failure;
-//  - Else the aggregated state is Idle
+//  - Else if at least one SubConn is Idle, the aggregated state is Idle;
+//  - Else there are no subconns and the aggregated state is Transient Failure
 //
 // Shutdown is not considered.
 func (cse *ConnectivityStateEvaluator) RecordTransition(oldState, newState connectivity.State) connectivity.State {
@@ -394,6 +396,8 @@ func (cse *ConnectivityStateEvaluator) RecordTransition(oldState, newState conne
 			cse.numConnecting += updateVal
 		case connectivity.TransientFailure:
 			cse.numTransientFailure += updateVal
+		case connectivity.Idle:
+			cse.numIdle += updateVal
 		}
 	}
 
@@ -407,5 +411,8 @@ func (cse *ConnectivityStateEvaluator) RecordTransition(oldState, newState conne
 	if cse.numTransientFailure > 0 {
 		return connectivity.TransientFailure
 	}
-	return connectivity.Idle
+	if cse.numIdle > 0 {
+		return connectivity.Idle
+	}
+	return connectivity.TransientFailure
 }

--- a/balancer/base/balancer.go
+++ b/balancer/base/balancer.go
@@ -214,10 +214,14 @@ func (b *baseBalancer) UpdateSubConnState(sc balancer.SubConn, state balancer.Su
 		}
 		return
 	}
-	if oldS == connectivity.TransientFailure && s == connectivity.Connecting {
-		// Once a subconn enters TRANSIENT_FAILURE, ignore subsequent
+	if oldS == connectivity.TransientFailure &&
+		(s == connectivity.Connecting || s == connectivity.Idle) {
+		// Once a subconn enters TRANSIENT_FAILURE, ignore subsequent IDLE or
 		// CONNECTING transitions to prevent the aggregated state from being
 		// always CONNECTING when many backends exist but are all down.
+		if s == connectivity.Idle {
+			sc.Connect()
+		}
 		return
 	}
 	b.scStates[sc] = s

--- a/balancer/base/balancer.go
+++ b/balancer/base/balancer.go
@@ -133,6 +133,7 @@ func (b *baseBalancer) UpdateClientConnState(s balancer.ClientConnState) error {
 			}
 			b.subConns[aNoAttrs] = subConnInfo{subConn: sc, attrs: a.Attributes}
 			b.scStates[sc] = connectivity.Idle
+			b.csEvltr.RecordTransition(connectivity.Shutdown, connectivity.Idle)
 			sc.Connect()
 		} else {
 			// Always update the subconn's address in case the attributes
@@ -242,7 +243,6 @@ func (b *baseBalancer) UpdateSubConnState(sc balancer.SubConn, state balancer.Su
 		b.state == connectivity.TransientFailure {
 		b.regeneratePicker()
 	}
-
 	b.cc.UpdateState(balancer.State{ConnectivityState: b.state, Picker: b.picker})
 }
 

--- a/balancer/roundrobin/roundrobin.go
+++ b/balancer/roundrobin/roundrobin.go
@@ -47,7 +47,7 @@ func init() {
 type rrPickerBuilder struct{}
 
 func (*rrPickerBuilder) Build(info base.PickerBuildInfo) balancer.Picker {
-	logger.Infof("roundrobinPicker: newPicker called with info: %v", info)
+	logger.Infof("roundrobinPicker: Build called with info: %v", info)
 	if len(info.ReadySCs) == 0 {
 		return base.NewErrPicker(balancer.ErrNoSubConnAvailable)
 	}

